### ISSUE TITLE
Can't export user data when data grows beyond a certain size

### ISSUE THREAD

**Issue by terrorist96** [EFForg/privacybadger#1318]
When clicking the download button, Chrome keeps saying Failed - Network error. I tried on a separate profile, and it works. I tried disabling all other extensions on my main profile and it still gives the error. Is there anything I can type into the console to figure out what is causing the issue?

![image](https://cloud.githubusercontent.com/assets/12802056/25060655/527dabf0-2170-11e7-9bd0-3035ed4d1cbc.png)

**Comment by ghostwords** [comment]
Are there any JavaScript errors or messages (before or after clicking Download) in the Developer Tools console for the options page?

![screenshot from 2017-04-17 09 50 42](https://cloud.githubusercontent.com/assets/794578/25090798/64f20a8a-2353-11e7-9a1e-1a4aceeeccf9.png)

**Comment by terrorist96** [comment]
Nope.
![image](https://cloud.githubusercontent.com/assets/12802056/25096007/412e6e72-236c-11e7-8286-1077584f5766.png)

**Comment by ghostwords** [comment]
My guess is that Badger's data can become [too large to fit in an anchor element](http://stackoverflow.com/a/37151835), our current download implementation.

What do you see if you run the following code in the options page console?
```javascript
chrome.storage.local.get(["action_map", "snitch_map", "settings_map"], function(maps) {
    var mapJSON = JSON.stringify(maps);
    var downloadURL = 'data:application/json;charset=utf-8,' + encodeURIComponent(mapJSON);
    console.log(downloadURL.length);
  });
```

**Comment by ghostwords** [comment]
Another problem with our current data download implementation is that instead of getting a file download in Firefox, we get the options page replaced by the data JSON. You can then save that data manually, but it's a worse user experience.

**Comment by terrorist96** [comment]
Running your code just returns `undefined` followed by the number `6577861`

**Comment by terrorist96** [comment]
Also, not too long ago, PB reported around 7k trackers in my options page. It now shows 36732. Did something change in the way it counts trackers?

**Comment by ghostwords** [comment]
That number is how long your data URL is, which comes out to something greater than 2 MB, which might be the current Chrome data URI limit.

**Comment by ghostwords** [comment]
This is the code that returns how many potential trackers your Badger got so far:
```javascript
function getOrigins() {
  var origins = {};
  var action_map = badger.storage.getBadgerStorageObject('action_map');
  for (var domain in action_map.getItemClones()) {
    var action = badger.storage.getBestAction(domain);
    // Do not show non tracking origins
    if(action != constants.NO_TRACKING){
      origins[domain] = action;
    }
  }
  return origins;
}
console.log(Object.keys(getOrigins()).length);
```
I don't think we changed this stuff recently.

When you import data, you combine, not overwrite. Would that explain the increase?

**Comment by terrorist96** [comment]
I haven't imported anything, so no. Just seems strange that it took a year or so to get to 7k and not too long after that, it jumped to 36k.

**Comment by ghostwords** [comment]
If you run the following in Badger's background page, what do you see? It should print some info about your top ten most popular (third-party) base domains. You might have thousands of like `googlevideo.com` subdomains, let's see.
```javascript
(function () {
  let action_map = badger.storage.getBadgerStorageObject('action_map').getItemClones();

  let result = _.chain(Object.keys(action_map))
    .countBy(domain => getBaseDomain(domain))
    .map((v, k) => {
      return {
        base: k,
        count: v
      }
    })
    .sortBy('count')
    .reverse()
    .slice(0, 10)
    .map(i => {
      i.domains = _.filter(Object.keys(action_map), d => d.endsWith(i.base));
      return i;
    })
    .value();

  console.log(JSON.stringify(result, null, 2));
}());
```

**Comment by terrorist96** [comment]
Here's the top one:
    "base": "pubnub.com",
    "count": 27318,
The rest don't come close to this. That looks to be the culprit. A whole lot of  `ps########.pubnub.com` subdomains.

**Comment by ghostwords** [comment]
Whoa, cool, looks like we got our answer. Do you mind sending me your snitch map for pubnub?
```javascript
(function () {
  const STR = "pubnub.com";
  console.log("**** SNITCH_MAP for", STR);
  _.each(badger.storage.getBadgerStorageObject('snitch_map').getItemClones(), (sites, domain) => {
    if (domain.indexOf(STR) != -1) console.log(domain, JSON.stringify(sites, null, 2));
  });
}());
```

**Comment by terrorist96** [comment]
Email.

**Comment by mirh** [comment]
> I haven't imported anything, so no. Just seems strange that it took a year or so to get to 7k and not too long after that, it jumped to 36k.

I second this wondering. 

Why do I have a 1.5MB export file.. When even in a year or two, I'll have manually overridden a dozen or two websites?

**Comment by ghostwords** [comment]
What are your most popular domains? To run the [above script](https://github.com/EFForg/privacybadger/issues/1318#issuecomment-297056299) in Privacy Badger's background page on Firefox:
1. Open the URL `about:debugging`
1. Enable add-on debugging and click Debug next to Privacy Badger
1. Click the OK button on the popup warning about remote debugging.
1. Enter the above script into the console after the >>

**Comment by mirh** [comment]
googlevideo.com it seems 😁

Now that I got how PB works though, just 1,5MB after 2 years seems a pretty good result considering how many pages I open per day.

**Comment by JDTm** [comment]
My list is 11436 Domains Long, could i remove some subdomains in chrome? Want to switch from chrome to google.

**Comment by ghostwords** [comment]
Fixed by cde3fd8f62c2de1044fc9e73a0f179962f93a9d7.
